### DISCUSSION TITLE
[Help]: Embla behaviour changes by device/screen size

### DISCUSSION THREAD

**Discussion by mk2moo** [davidjerleke/embla-carousel#358] [Help] [answered]
Hello,

I'm finding embla is changing behaviours by device and screen size on my website [ad-bank.co.uk](https://ad-bank.co.uk/all-ads/?content=carousel).

By behaviour I'm referring to:

Sometimes snapping, sometimes not
Sometimes, centered, sometimes not
Sometimes nav buttons work to correct slide, sometimes skips them all

This is behaviour which was not present before. I have attached a screen recording of the not working part.

In terms of set up I have the following:

This is fired last on page to ensure all other JS and css is loaded before.

This is making use of the multi carousel example.

Overall, any direction on why this behaviour would be really appreciated.

<script src="https://unpkg.com/embla-carousel/embla-carousel.umd.js"></script>
```
const setupPrevNextBtns = (prevBtn, nextBtn, embla) => {
  prevBtn.addEventListener("click", embla.scrollPrev, false);
  nextBtn.addEventListener("click", embla.scrollNext, false);
};

const disablePrevNextBtns = (prevBtn, nextBtn, embla) => {
  return () => {
    if (embla.canScrollPrev()) prevBtn.removeAttribute("disabled");
    else prevBtn.setAttribute("disabled", "disabled");

    if (embla.canScrollNext()) nextBtn.removeAttribute("disabled");
    else nextBtn.setAttribute("disabled", "disabled");
  };
};
const setupDotBtns = (dotsArray, embla) => {
  dotsArray.forEach((dotNode, i) => {
    dotNode.classList.add("embla__dot");
    dotNode.addEventListener("click", () => embla.scrollTo(i), false);
  });
};

const generateDotBtns = (dots, embla) => {
  const scrollSnaps = embla.scrollSnapList();
  const dotsFrag = document.createDocumentFragment();
  const dotsArray = scrollSnaps.map(() => document.createElement("button"));
  dotsArray.forEach((dotNode) => dotsFrag.appendChild(dotNode));
  dots.appendChild(dotsFrag);
  return dotsArray;
};

const selectDotBtn = (dotsArray, embla) => () => {
  const previous = embla.previousScrollSnap();
  const selected = embla.selectedScrollSnap();
  dotsArray[previous].classList.remove("is-selected");
  dotsArray[selected].classList.add("is-selected");
};

const setupEmblaCarousel = (emblaNode, options) => {
  const viewPort = emblaNode.querySelector(".embla__viewport");
  const prevBtn = emblaNode.querySelector(".embla__button--prev");
  const nextBtn = emblaNode.querySelector(".embla__button--next");
  const dots = emblaNode.querySelector(".embla__dots");
  const embla = EmblaCarousel(viewPort, options);
  const dotsArray = generateDotBtns(dots, embla);
  const setSelectedDotBtn = selectDotBtn(dotsArray, embla);
  const disablePrevAndNextBtns = disablePrevNextBtns(prevBtn, nextBtn, embla);

  setupPrevNextBtns(prevBtn, nextBtn, embla);
  setupDotBtns(dotsArray, embla);

  embla.on("select", setSelectedDotBtn);
  embla.on("select", disablePrevAndNextBtns);
  embla.on("init", setSelectedDotBtn);
  embla.on("init", disablePrevAndNextBtns);
};

const options = { loop: false, align: "center", containScroll: "keepSnaps" };
const emblaNodes = [].slice.call(document.querySelectorAll(".embla"));
const emblaCarousels = emblaNodes.map((emblaNode) =>
  setupEmblaCarousel(emblaNode, options)
);
```

https://user-images.githubusercontent.com/22149645/186144653-d2c8c4cf-12c7-4721-b395-faf8883ad0aa.mp4

**Comment by davidjerleke** [comment, marked as answer]
Hi @mk2moo,

Thank you for your question. You need to add the following to your code for buttons to work correctly, because Embla will trigger the `reInit` event on rezise.

```diff
const setupEmblaCarousel = (emblaNode, options) => {
  const viewPort = emblaNode.querySelector(".embla__viewport");
  const prevBtn = emblaNode.querySelector(".embla__button--prev");
  const nextBtn = emblaNode.querySelector(".embla__button--next");
  const dots = emblaNode.querySelector(".embla__dots");
  const embla = EmblaCarousel(viewPort, options);
  const dotsArray = generateDotBtns(dots, embla);
  const setSelectedDotBtn = selectDotBtn(dotsArray, embla);
  const disablePrevAndNextBtns = disablePrevNextBtns(prevBtn, nextBtn, embla);

  setupPrevNextBtns(prevBtn, nextBtn, embla);
  setupDotBtns(dotsArray, embla);

  embla.on("select", setSelectedDotBtn);
  embla.on("select", disablePrevAndNextBtns);
  embla.on("init", setSelectedDotBtn);
  embla.on("init", disablePrevAndNextBtns);
+ embla.on("reInit", setSelectedDotBtn);
+ embla.on("reInit", disablePrevAndNextBtns);
+ 
+ return embla
};

const options = { loop: false, align: "center", containScroll: "keepSnaps" };
const emblaNodes = [].slice.call(document.querySelectorAll(".embla"));
- const emblaCarousels = emblaNodes.map((emblaNode) =>
-  setupEmblaCarousel(emblaNode, options)
- );

+ const emblaCarousels = emblaNodes.map((emblaNode) => setupEmblaCarousel(emblaNode, options));
```

As for the snapping, I would like to know if you use any JS library for your layout? Because when loading/reloading the page there's a brief moment where the columns have a different size before they jump into position and get smaller. That's what's causing Embla to pick up the wrong dimensions. When resizing the window the carousel snaps into correct position. Either get rid of that layout thrashing with columns changing size, or run `embla.reInit()` on all your carousel instances when the jump is over.

I hope this helps you on the right track.

Best,
David

**Reply by mk2moo** [reply]
Amazing, thank you for the direction. Yeah colcade is a masonry layout, I'm still waiting on the grid masonry to be impemented.

The embla script is placed after the colcade which I thought would cover the issue.

It actually worked perfectly fine for 12 months and only until the last week the issue has cropped up.

Your direction on the resize is correct as when I adjust between screen width, it aligns correctly.

Sorry to be a pain but any direction on implementing reInit is appreciated.

**Reply by davidjerleke** [reply]
Hello again @mk2moo,

### The easy solution
> It actually worked perfectly fine for 12 months and only until the last week the issue has cropped up.

The easy solution to this problem is to change the following script src:

```html 
<script src="https://unpkg.com/embla-carousel/embla-carousel.umd.js"></script>
```

...into:

```html 
<script src="https://unpkg.com/embla-carousel@6.2.0/embla-carousel.umd.js"></script>
```

Using the following src `https://unpkg.com/embla-carousel/embla-carousel.umd.js` will always grab the latest release. And when a new major version is released (example: `6.x.x --> 7.x.x`), there will be breaking changes. Your problem is caused by breaking changes because Embla version 7 is using `px` as transform unit instead of `%`. `%` will adapt even when the sizes of elements change but `px` always need to be re-calculated.

⚠️ Please note that this solution won't enable you to use [v7 features](https://github.com/davidjerleke/embla-carousel/releases/tag/v7.0.0) and up.

### The more future proof solution
> The embla script is placed after the colcade which I thought would cover the issue.

This isn't always sufficient. Especially if the loaded scripts have to be initialized manually using a `constructor` like Embla:

```ts
const embla = EmblaCarousel(viewPort, options);
``` 

...and Colcade:

```ts
const colc = new Colcade(grid, {
  columns: '.grid-col',
  items: '.grid-item'
});
```

In these cases, what matters is the order of which the `constructors` are invoked (assuming the constructors aren't doing async things):

```ts
// init Colcade first
const colc = new Colcade(grid, { 
  columns: '.grid-col',
  items: '.grid-item'
});

// ...and Embla after
const embla = EmblaCarousel(viewPort, options); 
```

This will enable you to use Embla v7 and up.

Let me know if it helps.

Best,
David

**Reply by mk2moo** [reply]
Thank you David. I really appreciate the direction.

I have gone back to 6.2 and adjusted the colcade init to be done via JS and before any loading of embla.

I'll work on getting future proofed in sandbox. All in, it seems to be behaving correctly now.

Again, big thanks.

**Reply by davidjerleke** [reply]
Glad to hear that @mk2moo. Please mark my response as the answer if it solved your problem. Thanks!
